src/legacy/core_plugins/data/public/search/aggs/buckets/lib/time_buckets/time_buckets.ts

```diff
@@ -0,0 +1,460 @@
+    const duration = this.getDuration();
+
+    // either pull the interval from state or calculate the auto-interval
+    const readInterval = () => {
+      const interval = this._i;
+      if (moment.isDuration(interval)) return interval;
+      return calcAutoIntervalNear(this.getConfig('histogram:barTarget'), Number(duration));
+    };
+
+    const parsedInterval = readInterval();
+
+    // check to see if the interval should be scaled, and scale it if so
+    const maybeScaleInterval = (interval: moment.Duration) => {
+      if (!this.hasBounds() || !duration) {
+        return interval;
+      }
+
+      const maxLength: number = this.getConfig('histogram:maxBars');
+      const approxLen = Number(duration) / Number(interval);
+
+      let scaled;
+
+      if (approxLen > maxLength) {
+        scaled = calcAutoIntervalLessThan(maxLength, Number(duration));
+      } else {
+        return interval;
+      }
+
+      if (+scaled === +interval) return interval;
+
+      interval = decorateInterval(interval);
+      return Object.assign(scaled, {
+        preScaled: interval,
+        scale: Number(interval) / Number(scaled),
+        scaled: true,
+      });
+    };
+
+    // append some TimeBuckets specific props to the interval
+    const decorateInterval = (interval: moment.Duration): TimeBucketsInterval => {
+      const esInterval = useNormalizedEsInterval
+        ? convertDurationToNormalizedEsInterval(interval)
+        : convertIntervalToEsInterval(String(this._originalInterval));
+      const prettyUnits = moment.normalizeUnits(esInterval.unit);
+
+      return Object.assign(interval, {
+        description:
+          esInterval.value === 1 ? prettyUnits : esInterval.value + ' ' + prettyUnits + 's',
+        esValue: esInterval.value,
+        esUnit: esInterval.unit,
+        expression: esInterval.expression,
+        overflow:
+          Number(duration) > Number(interval)
+            ? moment.duration(Number(interval) - Number(duration))
+            : false,
+      });
+    };
+
+    if (useNormalizedEsInterval) {
+      return decorateInterval(maybeScaleInterval(parsedInterval));
+    } else {
+      return decorateInterval(parsedInterval);
+    }
+  }
+
+  /**
+   * Get a date format string that will represent dates that
+   * progress at our interval.
+   *
+   * Since our interval can be as small as 1ms, the default
+   * date format is usually way too much. with `dateFormat:scaled`
+   * users can modify how dates are formatted within series
+   * produced by TimeBuckets
+   *
+   * @return {string}
+   */
+  getScaledDateFormat() {
+    const interval = this.getInterval();
+    const rules = this.getConfig('dateFormat:scaled');
+
+    for (let i = rules.length - 1; i >= 0; i--) {
+      const rule = rules[i];
+      if (!rule[0] || (interval && interval >= moment.duration(rule[0]))) {
+        return rule[1];
+      }
+    }
+
+    return this.getConfig('dateFormat');
+  }
+
+  getScaledDateFormatter() {
```

Comment:
inline inside date_histogram .. i think thats the only place its used. everything else should use date_histograms `agg.getFormat()`